### ISSUE TITLE
Port twisted.python.roots to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8131]
|<img alt="jrolli's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jrolli reported|
|-|-|
|Trac ID|trac#8131|
|Type|enhancement|
|Created|2015-11-29 06:35:10Z|
|Branch|https://github.com/twisted/twisted/tree/roots-py3-8131|

twisted.python.roots cannot be imported in Python 3 because it relies on types.InstanceType

Attachments:

* [twisted_python_roots_py3-8131.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f4f/f4f23b3d1c967d4a197869692084d155751854ff/8e66e024c73c7224585fbdb64d29c4a8c4da2cc9.patch) (2095 bytes) - added by jrolli on 2015-11-29 06:45:51Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8131 8131
type__enhancement enhancement
reporter__jrolli jrolli
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_roots_py3_8131 branches/roots-py3-8131
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1448778910331130 1448778910331130
changetime__1449031562890193 1449031562890193
version__None None
owner__None None

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jrolli's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|jrolli commented|
|-|-|

In order to allow unit testing in Python 3, the twisted.test.test_roots needs to be changed because it relies on types.IntType

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

(In [46368]) Branching to roots-py3-8131.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

(In [46369]) Apply patch from jrolli, refs [#8131](https://github.com/twisted/twisted/issues/8131)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl set status to `closed`|
|-|-|

(In [46378]) Merge roots-py3-8131: Port twisted.python.roots to Python 3.

Author: jrolli
Reviewer: hawkowl
Fixes: [#8131](https://github.com/twisted/twisted/issues/8131)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

Hi jrolli, thanks for this. Just for future reference, we have some further information about Python 3 porting at https://twistedmatrix.com/trac/wiki/Plan/Python3 -- the only relevant part here was a .write() not using a bytestring (but that was fine) and the lack of ```__future__``` imports. I fixed those up before merging, so it's all good - but for future reference it's important to keep in mind :)

Thanks again!
